### PR TITLE
[CIS-1459] Make Data Formatters Configurable

### PR DESCRIPTION
**Pull request by nuno-vieira**
### 🔗 Issue Link
CIS-1459
https://github.com/GetStream/stream-chat-swift/issues/1739

### 🎯 Goal
Provide a way to change the data formatting in the SDK.

### 🛠 Implementation
Expose all formatters in the `Appearance` configuration through the `Appearance.default.formatters`. Each formatter is represented by a Protocol so that the customer can provide their own implementation or change the default ones.

### 🎨 Changes
N/A

### 🧪 Testing

Example on how to change a default formatter:
```swift
let defaultTimestampFormatter = DefaultMessageTimestampFormatter()
defaultTimestampFormatter.dateFormatter.timeStyle = .medium
Appearance.default.formatters.messageTimestamp = defaultTimestampFormatter
```

Example on how to provide a custom formatter:
```swift
class CustomUserLastActivityFormatter: UserLastActivityFormatter {
    public lazy var dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .short
        formatter.doesRelativeDateFormatting = true
        formatter.locale = .autoupdatingCurrent
        return formatter
    }()

    func format(_ date: Date) -> String? {
        "last active: " + dateFormatter.string(from: date)
    }
}
...
Appearance.default.formatters.userLastActivity = CustomUserLastActivityFormatter()
```

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] Changelog is updated with client-facing changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 # Upcoming
 
+### ✅ Added
+- Make Date Formatters mutable to allow customization [#1742](https://github.com/GetStream/stream-chat-swift/pull/1742)
```
Comment (tbarbugli) [inline CHANGELOG.md:7]: Make date formatters configurable sounds better ;)

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift [hunk 2]
```diff
@@ -57,7 +57,7 @@ open class ChatMessageContentView: _View, ThemeProvider {
     }
 
     /// The date formatter of the `timestampLabel`
-    public lazy var dateFormatter: DateFormatter = .makeDefault()
+    public lazy var dateFormatter: DateFormatter = .messageTimestamp
```
Comment (tbarbugli) [inline Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:60]: doc here does not make a lot of sense

**Review by tbarbugli — CHANGES_REQUESTED** [review]
We should add something to the SDK docs about this

**Review by polqf — APPROVED** [review]

**Comment by nuno-vieira** [comment]
> We should add something to the SDK docs about this

Will add the docs once we define the final API 👍

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:60]: I've updated them a little bit now, but it actually makes sense, this is the formatted used by the `timestampLabel` in this case. Were you thinking of a better doc?

File: Sources/StreamChatUI/Utils/DateFormatter+Extensions.swift [hunk 3]
```diff
@@ -5,33 +5,41 @@
...
     
-    /// Formatter that is used to format date for scrolling overlay that should display
-    /// day when message below was sent
-    public static let messageListDateOverlay: DateFormatter = {
-        let df = DateFormatter()
-        df.setLocalizedDateFormatFromTemplate("MMMdd")
-        df.locale = .autoupdatingCurrent
-        return df
+    /// A formatter that converts the message date to textual representation and used as a top overlay in the message list.
+    public static var messageListDateOverlay: DateFormatter = {
```
Comment (evsaev) [inline Sources/StreamChatUI/Utils/DateFormatter+Extensions.swift:18]: The approach we'd like to follow is:
```swift
struct Appearance {
  var dateFormatters = DateFormatters()

  struct DateFormatters {
     var formatter1 = …
     var formatter2 = …
  }
}
```

The **messageListDateOverlay** is the only publicly exposed static formatter (which we have to maintain) that is in conflict with the `Appearance`, right? 🤔 What if we make it forwarding to `Appearance.default`:
```swift
extension DateFormatter {
    @available(*, deprecated, message: "Please use `Appearance.default.dateFormatters.formatter1` instead")
    public static var messageListDateOverlay: DateFormatter {
      get { Appearance.default.dateFormatters.formatter1 }
      set { Appearance.default.dateFormatters.formatter1 = newValue }
    }
}
```

It's fine formatters are not static thanks to `Appearance.default` used in 99% of cases. A shared instance get instantiated with all its props just once when first accessed. And it's still possible to customize the formatter for the specific view-hierarchy subtree.
 
@nuno-vieira  WDYT?

Comment (evsaev) [reply Sources/StreamChatUI/Utils/DateFormatter+Extensions.swift:18]: we can also make dateFormatters and all underlying pros `lazy`

Comment (nuno-vieira) [reply Sources/StreamChatUI/Utils/DateFormatter+Extensions.swift:18]: We also have Number formatters. I will change the API and try to change to `Appearance.default.formatters` 👍

Comment (nuno-vieira) [reply Sources/StreamChatUI/Utils/DateFormatter+Extensions.swift]: ✅

**Review by bielikb — APPROVED** [review]

**Comment by nuno-vieira** [comment]
Converted to draft, since there will be some team discussions on the API

File: docusaurus/docs/iOS/uikit/data-formatting.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
+---
+title: Data Formatting
+---
+
+import ThemingNote from '../common-content/theming-note.md'
+
+You can customize how data is formatted across all the UI components provided by `StreamChatUI`. The SDK allows you to change the data formatting through the `Appearance` configuration where you can change, for example, how message dates, video durations, and times are formatted in each component. Each formatter can be accessed through the `Appearance.formatters` property. You can replace a formatter with a custom one since each formatter is represented by a `Protocol`, but you can also do simple changes by changing the default formatters provided by the SDK.
```
Comment (bielikb) [inline docusaurus/docs/iOS/uikit/data-formatting.md:7]: This paragraph could be rephrased. eg

```suggestion
You can customize how data is formatted across all the UI components provided by `StreamChatUI`. The SDK allows you to change the data formatting through the `Appearance.formatters` configuration.

Following data can be formatted:
* channel name
* message timestamps
* video durations & times

For simple customizations, use the default formatters provided by the SDK. Furtherly you can leverage your own formatter by adhering it to the respective data formatter `protocol`. 
```

File: docusaurus/docs/iOS/uikit/data-formatting.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        return formatter
+    }()
+
+    func format(_ date: Date) -> String? {
+        "last active: " + dateFormatter.string(from: date)
+    }
+}
+```
+
+Then, set the custom formatter in the `Appearance` config:
```
Comment (bielikb) [inline docusaurus/docs/iOS/uikit/data-formatting.md:47]:
```suggestion
Finally, set the custom formatter in the `Appearance` config:
```

**Review by bielikb — COMMENTED** [review]
Great job @nuno-vieira. I left two doc suggestions

File: Sources/StreamChatUI/Appearance+Formatters/MessageDateSeparatorFormatter.swift [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+/// A formatter that converts the message date separator used in the message list to textual representation.
+/// This formatter is used to display the message date between each group of messages
+/// and the top date overlay in the message list.
+public protocol MessageDateSeparatorFormatter {
+    func format(_ date: Date) -> String
+}
+
+/// The default message date separator formatter.
+open class DefaultMessageDateSeparatorFormatter: MessageDateSeparatorFormatter {
+    public lazy var dateFormatter: DateFormatter = {
```
Comment (polqf) [inline Sources/StreamChatUI/Appearance+Formatters/MessageDateSeparatorFormatter.swift:16]: We need to be careful with that. `lazy` vars are not thread safe.

Comment (polqf) [reply Sources/StreamChatUI/Appearance+Formatters/MessageDateSeparatorFormatter.swift:16]: I'm seeing that this was already the case 🤔. We might need to review in the future lazy implementations that are accessed globally.

**Review by polqf — APPROVED** [review]

File: Sources/StreamChatUI/Appearance+Formatters/MessageDateSeparatorFormatter.swift (hunk 6 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/Appearance+Formatters/MessageDateSeparatorFormatter.swift:16]: That makes sense 👍 I can easily change this to `var`, will do it now 👍 DateFromatter's are lazily created either way only when a function is called (like `string(from:)`), so it is fine to have them has normal vars in this case.

File: docusaurus/docs/iOS/uikit/data-formatting.md (hunk 4 above)
Comment (nuno-vieira) [reply docusaurus/docs/iOS/uikit/data-formatting.md]: @bielikb I've simplified this introduction, let me know if you like it 🙂

File: Sources/StreamChatUI/Appearance+Formatters/MessageDateSeparatorFormatter.swift (hunk 6 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/Appearance+Formatters/MessageDateSeparatorFormatter.swift]: @polqf Done ✅

**Review by bielikb — APPROVED** [review]

**Pull request merged**